### PR TITLE
Blog App - processing data in models

### PR DESCRIPTION
**Pull request by EidHachem**
## In this PR I implemented the below:

- Created model classes for all entities
- Set up associations between models.
- Created custom methods as below:
    - Users
        A method that returns the 3 most recent posts for a given user.
    - Posts
        A method that updates the posts counter for a user.
        A method that returns the 5 most recent comments for a given post.
    - Comments
        A method that updates the comments counter for a post.
    - Likes
        A method that updates the likes counter for a post.

### CODE REVIEW

File: app/models/post.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class Post < ApplicationRecord
+  belongs_to :author, class_name: 'User'
+  has_many :likes
+  has_many :comments
+
+  def update_counter
+    author.update(post_counter: author.posts.count)
+  end
```
Comment (OLIPLICHE) [inline app/models/post.rb]: ## Optional:
- Please you can use one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - `after_save` to call the `#update_post_counter` method every time a new post has been saved to database, then make the `update_post_counter` method private. Example:
```ruby
 after_save :update_post_counter

  private
```

```ruby
def update_post_counter
    author.increment!(:posts_counter)
  end
```

File: app/models/comment.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+class Comment < ApplicationRecord
+  belongs_to :author, class_name: 'User'
+  belongs_to :post
+
+  def update_comments_counter
+    post.update(comments_counter: post.comments.count)
+  end
```
Comment (OLIPLICHE) [inline app/models/comment.rb:11]: ## Optional:
- Here too, you can use one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - `after_save` to call the `#update_comment_counter` method every time a new post has been saved to database, then make the `update_comments_counter` method private

File: app/models/like.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+class Like < ApplicationRecord
+  belongs_to :author, class_name: 'User'
+  belongs_to :post
+
+  def update_likes_counter
+    post.update(likes_counter: post.likes.count)
+  end
```
Comment (OLIPLICHE) [inline app/models/like.rb:11]: ## Optional
- Same here; `after_save` `:update_likes_counter`

**Review by OLIPLICHE — APPROVED** [review]
Hi @EidHachem , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Excellent 👍
- The requirements were followed and applied accordingly 👍

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by EidHachem** [comment]
Thank you so much @OLIPLICHE, that was really helpful.

**Pull request merged**